[dtensor][9/n] matrix ops to generate strategy

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111278
* #111234
* __->__ #110717
* #109145

This PR switches matrix ops to generate the sharding strategies, and
with the cost selection algorithm introduced in the previous PR we are
able to enable this and more ops to leverage strategy based sharding
prop

This also fixes a bunch of corner cases that existing propagation does
not cover, resulting in full coverage for baddbmm

wanchaol committed:
[dtensor][9/n] matrix ops to generate strategy

This PR switches matrix ops to generate the sharding strategies, and
with the cost selection algorithm introduced in the previous PR we are
able to enable this and more ops to leverage strategy based sharding
prop

This also fixes a bunch of corner cases that existing propagation does
not cover, resulting in full coverage for baddbmm

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][9/n] matrix ops to generate strategy"


This PR switches matrix ops to generate the sharding strategies, and
with the cost selection algorithm introduced in the previous PR we are
able to enable this and more ops to leverage strategy based sharding
prop

This also fixes a bunch of corner cases that existing propagation does
not cover, resulting in full coverage for baddbmm

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][9/n] matrix ops to generate strategy"


This PR switches matrix ops to generate the sharding strategies, and
with the cost selection algorithm introduced in the previous PR we are
able to enable this and more ops to leverage strategy based sharding
prop

This also fixes a bunch of corner cases that existing propagation does
not cover, resulting in full coverage for baddbmm

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][9/n] matrix ops to generate strategy"


This PR switches matrix ops to generate the sharding strategies, and
with the cost selection algorithm introduced in the previous PR we are
able to enable this and more ops to leverage strategy based sharding
prop

This also fixes a bunch of corner cases that existing propagation does
not cover, resulting in full coverage for baddbmm

[ghstack-poisoned]

fduwjj commented on test/distributed/_tensor/test_matrix_ops.py:
```diff
@@ -219,25 +202,11 @@ def test_placement_comb(
-            # tests that currently pass
-            for spec in passlist:
+            # test all combos
+            for spec in shard_specs_comb:
```
nice!

fduwjj commented on torch/distributed/_tensor/ops/matrix_ops.py:
```diff
@@ -1,128 +1,184 @@
+    # infer the broadcast dims map, where it maps from the common shape dim to the input shape dim
+    # this is aligned with the broadcast semantics
+    common_ndim = len(common_shape)
+    input_ndim = len(input_shape)
```
nit: maybe we want to have a check on the length is equal?

fduwjj replied:
NVM, looks like input_ndim <= common_ndim according to https://pytorch.org/docs/stable/notes/broadcasting.html.

fduwjj reviewed (approved):
The change and logic looks make sense to me.

wanchaol committed:
Update on "[dtensor][9/n] matrix ops to generate strategy"


This PR switches matrix ops to generate the sharding strategies, and
with the cost selection algorithm introduced in the previous PR we are
able to enable this and more ops to leverage strategy based sharding
prop

This also fixes a bunch of corner cases that existing propagation does
not cover, resulting in full coverage for baddbmm

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][9/n] matrix ops to generate strategy"


This PR switches matrix ops to generate the sharding strategies, and
with the cost selection algorithm introduced in the previous PR we are
able to enable this and more ops to leverage strategy based sharding
prop

This also fixes a bunch of corner cases that existing propagation does
not cover, resulting in full coverage for baddbmm

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][9/n] matrix ops to generate strategy"


This PR switches matrix ops to generate the sharding strategies, and
with the cost selection algorithm introduced in the previous PR we are
able to enable this and more ops to leverage strategy based sharding
prop

This also fixes a bunch of corner cases that existing propagation does
not cover, resulting in full coverage for baddbmm

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][9/n] matrix ops to generate strategy"


This PR switches matrix ops to generate the sharding strategies, and
with the cost selection algorithm introduced in the previous PR we are
able to enable this and more ops to leverage strategy based sharding
prop

This also fixes a bunch of corner cases that existing propagation does
not cover, resulting in full coverage for baddbmm

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][9/n] matrix ops to generate strategy"


This PR switches matrix ops to generate the sharding strategies, and
with the cost selection algorithm introduced in the previous PR we are
able to enable this and more ops to leverage strategy based sharding
prop

This also fixes a bunch of corner cases that existing propagation does
not cover, resulting in full coverage for baddbmm

[ghstack-poisoned]
